OpenRA.Mods.Cnc/Traits/ConyardChronoReturn.cs

```diff
@@ -0,0 +1,251 @@
+	{
+		[Desc("Sequence name with the baked-in vortex animation"), SequenceReference]
+		public readonly string Sequence = "pdox";
+
+		[Desc("Sprite body to play the vortex animation on.")]
+		public readonly string Body = "body";
+
+		[GrantedConditionReference]
+		[Desc("Condition to grant while the vortex animation plays.")]
+		public readonly string Condition = null;
+
+		[Desc("Amount of damage to apply each tick while the vortex animation plays.")]
+		public readonly int Damage = 1000;
+
+		[Desc("Apply the damage using these damagetypes.")]
+		public readonly HashSet<string> DamageTypes = new HashSet<string>();
+
+		[Desc("Actor to transform into when the timer expires during (un)deploy."), ActorReference, FieldLoader.Require]
+		public readonly string OriginalActor = null;
+
+		[Desc("Facing of the returned actor.")]
+		public readonly int Facing = 96;
+
+		public readonly string ChronoshiftSound = "chrono2.aud";
+
+		[Desc("The color the bar of the 'return-to-origin' logic has.")]
+		public readonly Color TimeBarColor = Color.White;
+
+		public object Create(ActorInitializer init) { return new ConyardChronoReturn(init, this); }
+	}
+
+	public class ConyardChronoReturn : INotifyCreated, ITick, ISync, ISelectionBar, IDeathActorInitModifier,
+		ITransformActorInitModifier, INotifyBuildComplete, INotifySold, INotifyTransform
+	{
+		readonly ConyardChronoReturnInfo info;
+		readonly WithSpriteBody wsb;
+		readonly Health health;
+		readonly Actor self;
+		readonly string faction;
+
+		IPositionable positionable;
+		ConditionManager conditionManager;
+		int conditionToken = ConditionManager.InvalidConditionToken;
+
+		Actor chronosphere;
+		int duration;
+		bool buildComplete;
+		bool selling;
+
+		[Sync]
+		int returnTicks = 0;
+
+		[Sync]
+		CPos origin;
+
+		[Sync]
+		bool triggered;
+
+		public ConyardChronoReturn(ActorInitializer init, ConyardChronoReturnInfo info)
+		{
+			this.info = info;
+			self = init.Self;
+
+			health = self.Trait<Health>();
+			wsb = self.TraitsImplementing<WithSpriteBody>().Single(w => w.Info.Name == info.Body);
+			faction = init.Contains<FactionInit>() ? init.Get<FactionInit, string>() : self.Owner.Faction.InternalName;
+
+			if (init.Contains<ChronoshiftReturnInit>())
+				returnTicks = init.Get<ChronoshiftReturnInit, int>();
+
+			if (init.Contains<ChronoshiftDurationInit>())
+				duration = init.Get<ChronoshiftDurationInit, int>();
+
+			if (init.Contains<ChronoshiftOriginInit>())
+				origin = init.Get<ChronoshiftOriginInit, CPos>();
+
+			if (init.Contains<ChronoshiftChronosphereInit>())
+				chronosphere = init.Get<ChronoshiftChronosphereInit, Actor>();
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+			positionable = self.TraitOrDefault<IPositionable>();
+		}
+
+		void TriggerVortex()
+		{
+			if (!string.IsNullOrEmpty(info.Condition))
```

Comment:
Do we want to check if `conditionToken == ConditionManager.InvalidConditionToken` here?